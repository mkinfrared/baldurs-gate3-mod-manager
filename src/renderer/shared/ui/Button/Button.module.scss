@import "../../styles/button-base.scss";

.Button {
  @extend %button-base;

  &.loading {
    .content {
      opacity: 0;
    }

    .spinner {
      opacity: 1;
      z-index: 1;
    }
  }

  .content {
    align-items: center;
    display: inline-block;
    gap: 1rem;
    max-height: var(--text-height);
    white-space: nowrap;
  }

  .spinner {
    opacity: 0;
    position: absolute;
    z-index: -1;
  }

  .content,
  .spinner {
    transition: opacity 250ms ease-in-out;
  }
}
