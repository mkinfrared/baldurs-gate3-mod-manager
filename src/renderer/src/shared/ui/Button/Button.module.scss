@import "../../styles/button-base";

%button {
  @extend %button-base;

  &.loading {
    .content {
      opacity: 0;
    }

    .spinner {
      opacity: 1;
      z-index: 1;
    }
  }

  .content {
    align-items: center;
    display: inline-flex;
    gap: 1rem;
    max-height: var(--text-height);
    white-space: nowrap;
  }

  .spinner {
    opacity: 0;
    position: absolute;
    z-index: -1;
  }

  .content,
  .spinner {
    transition: opacity 250ms ease-in-out;
  }
}

.Button {
  @extend %button;
}
