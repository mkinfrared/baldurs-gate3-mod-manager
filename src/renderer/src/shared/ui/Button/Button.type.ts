import { ButtonHTMLAttributes, ReactNode } from "react";

type ButtonAttributes = ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = ButtonAttributes & {
  /**
   * The children prop specifies the child elements or components.
   */
  children?: ReactNode;

  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;

  /**
   * The color of a button.
   */
  color?: ButtonColors;

  /**
   * Represents the loading state of a variable.
   */
  loading?: boolean;

  /**
   * Represents the visual variant of a button.
   */
  variant?: ButtonVariants;
};

export type ButtonColors = "error" | "primary" | "secondary" | "success";

export type ButtonVariants = "contained" | "outlined" | "transparent";
