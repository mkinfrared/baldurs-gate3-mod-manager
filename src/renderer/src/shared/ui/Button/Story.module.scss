%button-story {
  > section {
    &:not(:last-child) {
      margin-bottom: 2rem;
    }
  }

  .sameWidthList {
    display: grid;
    grid-auto-columns: 1fr;
    grid-auto-flow: column;
  }
}

.Story {
  @extend %button-story;

  .buttonBlock {
    display: flex;
    gap: 20px;
    margin-bottom: 20px;
  }

  .title {
    margin-block-end: 20px;
  }

  .subtitle {
    margin-block-end: 10px;
  }
}
