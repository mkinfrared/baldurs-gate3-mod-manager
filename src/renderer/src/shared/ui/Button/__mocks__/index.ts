export { default } from "./Button.mock";
