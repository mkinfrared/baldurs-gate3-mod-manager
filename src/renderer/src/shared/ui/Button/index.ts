export { Button } from "./Button";

export * from "./Button.type";
