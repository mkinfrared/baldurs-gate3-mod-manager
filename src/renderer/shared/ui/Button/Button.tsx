/* eslint-disable react/button-has-type */
import { classNames } from "renderer/shared/lib/helpers";

import css from "./Button.module.scss";
import { ButtonColors, ButtonProps, ButtonVariants } from "./Button.type";

/*
  renders an html button

  @example
  see Button.stories.tsx
 */
const Button = ({
  buttonRef,
  children,
  className,
  color = "primary",
  disabled = false,
  type = "button",
  variant = "contained",
  ...rest
}: ButtonProps) => {
  const colors: Record<ButtonColors, string> = {
    primary: css.primary,
    secondary: css.secondary,
    success: css.success,
    error: css.error,
  };

  const variants: Record<ButtonVariants, string> = {
    contained: css.contained,
    transparent: css.transparent,
  };

  const getButtonColor = (key: ButtonColors) => colors[key];
  const getButtonVariant = (key: ButtonVariants) => variants[key];

  const classes = [
    css.Button,
    className,
    getButtonVariant(variant),
    getButtonColor(color),
  ];

  return (
    <button
      className={classNames(...classes)}
      disabled={disabled}
      data-testid="Button"
      ref={buttonRef}
      type={type}
      {...rest}
    >
      <span className={css.content}>{children}</span>
    </button>
  );
};

export default Button;
