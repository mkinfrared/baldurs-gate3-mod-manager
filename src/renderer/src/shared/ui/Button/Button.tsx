/* eslint-disable react/button-has-type */
import { forwardRef } from "react";

import { classNames } from "@renderer/shared/lib/helpers";

import Spinner from "../Spinner";

import css from "./Button.module.scss";
import { ButtonColors, ButtonProps, ButtonVariants } from "./Button.type";

/**
 * Button component
 *
 * This component renders a customizable button that can have different
 * colors and variants, such as 'contained', 'transparent', and 'outlined'.
 * It also allows for child elements to be placed inside the button.
 * This button can also show a loading state where a Spinner is shown.
 * @example
 * // Renders a primary contained button with "Click me" as the label
 * <Button color="primary" variant="contained">Click me</Button>
 *
 * // Renders a disabled secondary outlined button with a loading spinner
 * <Button color="secondary" variant="outlined" isDisabled loading />
 */
const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      children,
      className,
      color = "primary",
      disabled = false,
      loading = false,
      type = "button",
      variant = "contained",
      ...rest
    },
    ref,
  ) => {
    const colors: Record<ButtonColors, string> = {
      primary: css.primary,
      secondary: css.secondary,
      success: css.success,
      error: css.error,
    };

    const variants: Record<ButtonVariants, string> = {
      contained: css.contained,
      transparent: css.transparent,
      outlined: css.outlined,
    };

    const getButtonColor = (key: ButtonColors) => colors[key];
    const getButtonVariant = (key: ButtonVariants) => variants[key];

    const classes = [
      css.Button,
      className,
      getButtonVariant(variant),
      getButtonColor(color),
    ];

    if (loading) {
      classes.push(css.loading);
    }

    return (
      <button
        className={classNames(...classes)}
        disabled={disabled}
        data-testid="Button"
        ref={ref}
        type={type}
        {...rest}
      >
        <Spinner className={css.spinner} type="bars" />
        <span className={css.content}>{children}</span>
      </button>
    );
  },
);

Button.displayName = "Button";

export { Button };
