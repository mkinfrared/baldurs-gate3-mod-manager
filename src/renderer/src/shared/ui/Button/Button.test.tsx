import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import Button from "./Button";

describe("<Button />", () => {
  const Component = <Button />;

  it("should be defined", () => {
    expect(Button).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("Button");

    expect(element).toBeDefined();
  });
});
