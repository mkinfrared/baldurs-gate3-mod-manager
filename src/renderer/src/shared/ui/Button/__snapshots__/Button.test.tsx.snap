// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Button /> > should match the snapshot 1`] = `
<div>
  <button
    class="Button contained primary"
    data-testid="Button"
    type="button"
  >
    <svg
      class="Spinner spinner"
    >
      bars.svg
    </svg>
    <span
      class="content"
    />
  </button>
</div>
`;
